### DISCUSSION TITLE
How to correctly apply metrics API in binary use case

### DISCUSSION THREAD

**Discussion by kapsner** [Lightning-AI/pytorch-lightning#6356] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule] [answered]
How would one correctly apply the Precision metric from v1.2.0 on, with the revised metrics api?

I am currently doing something like this:

```python
import torch
from pytorch_lightning import metrics

# example data
preds = [0] * 200 + [1] * 30 + [0] * 10 + [1] * 20
targets = [0] * 200 + [1] * 30 + [1] * 10 + [0] * 20

preds = torch.tensor(preds)
targets = torch.tensor(targets)

# define method for printing metrics
def _print_some_metrics(preds, targets, num_classes):
    precision = metrics.classification.Precision(
        num_classes=None, is_multiclass=False)
    recall = metrics.classification.Recall(
        num_classes=None, is_multiclass=False)
    f1 = metrics.classification.F1(num_classes=num_classes)
    f1beta = metrics.classification.FBeta(
        num_classes=num_classes,
        beta=2
    )

    accuracy = metrics.classification.Accuracy()
    avg_precision = metrics.classification.AveragePrecision(
        num_classes=None)
    confusion_matrix = metrics.ConfusionMatrix(num_classes=2)

    # print results
    print("Precision:\n{}\n".format(precision(preds, targets)))
    print("Recall:\n{}\n".format(recall(preds, targets)))
    print("F1:\n{}\n".format(f1(preds, targets)))
    print("F1-Beta:\n{}\n".format(f1beta(preds, targets)))

    print("AVG Precision:\n{}\n".format(avg_precision(preds, targets)))
    print("Accuracy:\n{}\n".format(accuracy(preds, targets)))
    print("ConfMat:\n{}\n".format(confusion_matrix(preds, targets)))

_print_some_metrics(preds, targets, num_classes=2)
```

Which gives me these results:
> Precision:
0.6000000238418579

> Recall:
0.75

> F1:
0.8846153616905212

> F1-Beta:
0.8846153616905212

> AVG Precision:
0.48846155405044556

> Accuracy:
0.8846153616905212

> ConfMat:
tensor([[200.,  20.],
        [ 10.,  30.]])

However, when calculating precision by hand [(TP / TP + FN)](https://en.wikipedia.org/wiki/Precision_and_recall) with the numbers from the contingency table, I get 30 / 50 = 0.6

Why does applying the precision class result in this (small) deviation?

Further, when logging the metrics on epoch_end steps inside my model, I am not able to reproduce the logged precision, recall or accuraccy numbers on the validation set with the output from the contingency table, logged on the same steps (I haven't validated the other metrics yet by hand).

It would be great to get some help on how to correctly apply the new metrics API for a binary use case.

**Comment by SkafteNicki** [comment, marked as answer]
for binary classification, where you are only interested in the positive class you should pass in `num_classes=1`. Here is your corrected code:
``` python
def _print_some_metrics(preds, targets, num_classes):
    precision = metrics.classification.Precision(
        num_classes=num_classes, is_multiclass=False)
    recall = metrics.classification.Recall(
        num_classes=num_classes, is_multiclass=False)
    f1 = metrics.classification.F1(num_classes=num_classes)
    f1beta = metrics.classification.FBeta(
        num_classes=num_classes,
        beta=2
    )

    accuracy = metrics.classification.Accuracy()
    avg_precision = metrics.classification.AveragePrecision(
        num_classes=num_classes)
    confusion_matrix = metrics.ConfusionMatrix(num_classes=2)

    # print results
    print("Precision:\n{}\n".format(precision(preds, targets)))
    print("Recall:\n{}\n".format(recall(preds, targets)))
    print("F1:\n{}\n".format(f1(preds, targets)))
    print("F1-Beta:\n{}\n".format(f1beta(preds, targets)))

    print("AVG Precision:\n{}\n".format(avg_precision(preds, targets)))
    print("Accuracy:\n{}\n".format(accuracy(preds, targets)))
    print("ConfMat:\n{}\n".format(confusion_matrix(preds, targets)))

_print_some_metrics(preds, targets, num_classes=1)
```
only `confusion_matrix` need to be set `num_classes=2` because you want the statistics for both classes.

**Reply by kapsner** [reply]
@SkafteNicki thx for your answer. IMHO setting `num_classes=1` doesn't make a change with this code and is further not mentioned anymore with regards to binary use cases in the new metrics API (>=v1.2.0) documentation.

I created a minimal working example to better explain what I mean. In this example I am comparing the metrics from the pytorch lightning class interface to those calculated "by hand" (colors indicate corresponding metrics):
![grafik](https://user-images.githubusercontent.com/44973495/110134130-5a332100-7dcd-11eb-8962-d788b442d8f7.png)

As you can see, the "comparison"-metrics, calculated with the numbers from the confusion matrix (see respective TP, FP, TN and FN values in columns H-K), have other values than those calculated by the pl-metrics API.

So my question is, if I am somehow applying the pl metrics API totally wrong, or if there might be a bug?

Here is the example code:
```python
import os
from numpy.lib.npyio import save

import torch
from torch.random import seed
from torch.utils.data import Dataset

from pl_examples import cli_lightning_logo
from pytorch_lightning import LightningModule, Trainer, seed_everything
from pytorch_lightning.loggers import CSVLogger

from pytorch_lightning import metrics
from pytorch_lightning.metrics.utils import to_categorical


class RandomDataset(Dataset):
    """
    >>> RandomDataset(size=10, length=20)  # doctest: +ELLIPSIS
    <...bug_report_model.RandomDataset object at ...>
    """

    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    """
    >>> BoringModel()  # doctest: +ELLIPSIS +NORMALIZE_WHITESPACE
    BoringModel(
      (layer): Linear(...)
    )
    """

    def __init__(self):
        """
        Testing PL Module

        Use as follows:
        - subclass
        - modify the behavior for what you want

        class TestModel(BaseTestModel):
            def training_step(...):
                # do your own thing

        or:

        model = BaseTestModel()
        model.training_epoch_end = None

        """
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def loss(self, batch, prediction):
        # An arbitrary loss to have a loss that updates the model weights during `Trainer.fit` calls
        return torch.nn.functional.mse_loss(prediction, torch.ones_like(prediction))

    def step(self, x):
        x = self.layer(x)
        out = torch.nn.functional.mse_loss(x, torch.ones_like(x))
        return out

    def training_step(self, batch, batch_idx):
        output = self.layer(batch)
        loss = self.loss(batch, output)
        return {"loss": loss}

    def training_step_end(self, training_step_outputs):
        return training_step_outputs

    def training_epoch_end(self, outputs) -> None:
        torch.stack([x["loss"] for x in outputs]).mean()

    def validation_step(self, batch, batch_idx):
        output = self.layer(batch)
        loss = self.loss(batch, output)
        return {"x": loss}

    def validation_epoch_end(self, outputs) -> None:
        torch.stack([x['x'] for x in outputs]).mean()

    def test_step(self, batch, batch_idx):
        output = self.layer(batch)
        loss = self.loss(batch, output)
        return {"y": loss}

    def test_epoch_end(self, outputs) -> None:
        torch.stack([x["y"] for x in outputs]).mean()

    def configure_optimizers(self):
        optimizer = torch.optim.SGD(self.layer.parameters(), lr=0.1)
        lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=1)
        return [optimizer], [lr_scheduler]


#  NOTE: If you are using a cmd line to run your script,
#  provide the cmd line as below.
#  opt = "--max_epochs 1 --limit_train_batches 1".split(" ")
#  parser = ArgumentParser()
#  args = parser.parse_args(opt)

def test_run():

    class TestModel(BoringModel):

        def __init__(self):
            super().__init__()

            self.num_classes = 2

            num_cls = None if self.num_classes == 2 else self.num_classes
            is_multiclass = False if num_cls is None else True

            self.valid_acc = metrics.classification.Accuracy(
                compute_on_step=False
            )

            self.valid_precision = metrics.classification.Precision(
                num_classes=num_cls,
                compute_on_step=False,
                is_multiclass=is_multiclass
            )
            self.valid_recall = metrics.classification.Recall(
                num_classes=num_cls,
                compute_on_step=False,
                is_multiclass=is_multiclass
            )

            self.valid_statscores = metrics.classification.StatScores(
                num_classes=num_cls,
                is_multiclass=is_multiclass
            )

            self._pl_metrics = [
                "valid_acc",
                "valid_precision",
                "valid_recall"
            ]

            self._compare_metrics = [
                "_accurcay",
                "_precision",
                "_recall"
            ]

        @staticmethod
        def _accurcay(tp, fp, tn, fn):
            return (tp + tn) / (tp + fp + tn + fn)

        @staticmethod
        def _precision(tp, fp, tn, fn):
            return tp / (tp + fp)

        @staticmethod
        def _recall(tp, fp, tn, fn):
            return tp / (tp + fn)

        @staticmethod
        def _y_to_cat(logits, targets):
            y_pred = to_categorical(
                tensor=torch.sigmoid(logits),
                argmax_dim=1
            )
            return y_pred, targets

        def validation_step(self, batch, batch_idx):
            output = self.layer(batch)
            loss = self.loss(batch, output)

            bs = torch.tensor([len(output)], dtype=torch.int16).type_as(output)

            # update metrics
            _preds, _targets = self._y_to_cat(
                logits=output,
                targets=torch.multinomial(
                    torch.tensor([.5, .5]),
                    len(output),
                    replacement=True
                )
            )

            for _metname in self._pl_metrics:
                eval("self." + _metname)(
                    preds=_preds,
                    target=_targets
                )

            self.valid_statscores(preds=_preds, target=_targets)

            return {"loss": loss, "batch_size": bs}

        def validation_epoch_end(self, outputs):
            # concat batch sizes
            batch_sizes = torch.stack(
                [x["batch_size"] for x in outputs]
            ).type_as(outputs[0]["loss"])

            # concat losses
            losses = torch.stack(
                [x["loss"] for x in outputs]
            ).type_as(outputs[0]["loss"])

            # calculating weighted mean loss
            avg_loss = torch.sum(losses * batch_sizes) / torch.sum(batch_sizes)

            self.log(
                name="loss/valid",
                value=avg_loss,
                prog_bar=False,
                logger=True,
                on_step=False,
                on_epoch=True
            )

            # compute metrics
            for _metname in self._pl_metrics:
                self.log(
                    name="pl/" + _metname,
                    value=eval("self." + _metname + ".compute()"),
                    prog_bar=False,
                    logger=True,
                    on_step=False,
                    on_epoch=True
                )

            tp, fp, tn, fn, _ = self.valid_statscores.compute()
            self.valid_statscores.reset()

            for _metname in self._compare_metrics:
                self.log(
                    name="comparison/" + _metname,
                    value=eval("self." + _metname + "(tp, fp, tn, fn)"),
                    prog_bar=False,
                    logger=True,
                    on_step=False,
                    on_epoch=True
                )

            for _m in ["tp", "fp", "tn", "fn"]:
                self.log(
                    name="comparison/" + _m,
                    value=eval(_m),
                    prog_bar=False,
                    logger=True,
                    on_step=False,
                    on_epoch=True
                )

    # fake data
    train_data = torch.utils.data.DataLoader(RandomDataset(32, 100))
    val_data = torch.utils.data.DataLoader(RandomDataset(32, 100))
    test_data = torch.utils.data.DataLoader(RandomDataset(32, 100))

    # logger
    csvlogger = CSVLogger(
        save_dir=os.getcwd()
    )
    # model
    seed_everything(0)
    model = TestModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        max_epochs=5,
        weights_summary=None,
        logger=csvlogger,
        deterministic=True
    )
    trainer.fit(model, train_data, val_data)
    trainer.test(test_dataloaders=test_data)


if __name__ == '__main__':
    cli_lightning_logo()
    test_run()
```

**Reply by kapsner** [reply]
@SkafteNicki `num_classes=1` still not gives the desired numbers... they are different from those calculated outside the metrics API. Here is a kaggle notebook with the code (version 1: numclasses = 1, version 2: numclasses = 2) https://www.kaggle.com/nonserial/notebooke3a0240813

**Reply by kapsner** [reply]
@SkafteNicki I can confirm that the metrics API from pl v.1.8.1 works as expected using `num_classes=1`.  Here is the corresponding code: https://www.kaggle.com/nonserial/fork-of-notebooke3a0240813

Could you please have a look at it? To me it seems as if there is a bug in the metrics API of at least pl v1.2.1...

**Reply by kapsner** [reply]
Since it is working as expected in v1.8.1, I have tranferred this discussion into a bug report: https://github.com/PyTorchLightning/pytorch-lightning/issues/6396

**Reply by SkafteNicki** [reply]
@kapsner please see my comment on the issue that you raised: https://github.com/PyTorchLightning/pytorch-lightning/issues/6396#issuecomment-792344545

**Reply by SkafteNicki** [reply]
Just reposting my answer here so anyone that finds this discussion can find the solution.

> Nothing is wrong with the metrics. I can get it working in v1.2 with the following changes to your code:
> 
>     1. initialize the metrics as the following:
> 
> 
> ```python
> self.valid_acc = metrics.classification.Accuracy()
> self.valid_precision = metrics.classification.Precision(num_classes=1, is_multiclass=False)
> self.valid_recall = metrics.classification.Recall(num_classes=1, is_multiclass=False)
> self.valid_statscores = metrics.classification.StatScores(num_classes=1, is_multiclass=False)
> ```
> 
>     2. remember to call `reset` after calling `compute`. Your code does this for the 
>        statscore metric but not for the other, which is the reason why you saw that the metrics began to diverge more and more
> 
> 
> ```python
> # compute metrics
> for _metname in self._pl_metrics:
>     self.log(
>         name="pl/" + _metname,
>         value=eval("self." + _metname + ".compute()"),
>         prog_bar=False,
>         logger=True,
>         on_step=False,
>         on_epoch=True
>     )
>     eval("self." + _metname + ".reset()") # this is missing
> ```
